File: Python/백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.py
from collections import deque

n, k = map(int, input().split())
belt = [0] + list(map(int, input().split()))
visit = [False] * (2*n + 1)
robot = deque()

start = 1
end = n
cnt = 0
res = 0

def moveBelt():
    global start, end
    start -= 1
    end -= 1
    if start < 1: start = 2*n
    if end < 1: end = 2*n

def moveRobot():
    global cnt
    for _ in range(len(robot)):
        cur = robot.popleft()
        visit[cur] = False

        if cur == end: continue

        nxt = cur + 1
        if nxt > 2*n:
            nxt = 1

        if belt[nxt] >= 1 and not visit[nxt]:
            belt[nxt] -= 1
            if belt[nxt] == 0: cnt += 1
            if nxt == end: continue

            visit[nxt] = True
            robot.append(nxt)

        else:
            visit[cur] = True
            robot.append(cur)

def makeRobot():
    global cnt
    if not visit[start] and belt[start] >= 1:
        visit[start] = True
        belt[start] -= 1
        robot.append(start)
        if belt[start] == 0:
            cnt += 1

while cnt < k:
    res += 1
    moveBelt()
    moveRobot()
    makeRobot()
print(res)

File: C++17/백준/Gold/20055. 컨베이어 벨트 위의 로봇/컨베이어 벨트 위의 로봇.cc
#include<iostream>
#include<vector>
#include<queue>
using namespace std;
int n, k, Start, End, res = 0, cnt = 0;
vector<int> belt;
vector<bool> visit;
queue<int> robot;
void MoveBelt();
void MoveRobot();
void AddRobot();

int main() {
	cin >> n >> k;
	Start = 1;
	End = n;
	belt.resize(2 * n + 1);
	visit.assign(2 * n + 1, false);

	for (int i = 1; i <= 2 * n; i++) cin >> belt[i];
	while (cnt < k){
		res++;
		MoveBelt();
		MoveRobot();
		AddRobot();
	}
	cout << res << "\n";
	return 0;
}
void MoveBelt() {
	Start--;
	End--;
	if (Start < 1) Start = 2 * n;
	if (End < 1) End = 2 * n;
}
void MoveRobot() {
	int Size = robot.size();
	for (int i = 0; i < Size; i++) {
		int cur = robot.front();
		visit[cur] = false;
		robot.pop();

		if (cur == End) continue;

		int next = cur + 1;
		if (next > 2 * n) next = 1;
		if (belt[next] > 0 && visit[next] == false) {
			belt[next]--;
			if (belt[next] == 0) cnt++;
			if (next == End) continue;
			visit[next] = true;
			robot.push(next);
		}
		else {
			visit[cur] = true;
			robot.push(cur);
		}
	}
}
void AddRobot() {
	if (visit[Start] == false && belt[Start] > 0) {
		visit[Start] = true;
		belt[Start]--;
		robot.push(Start);
		if (belt[Start] == 0) cnt++;
	}
}